### PR TITLE
Avoid collision w std macros use enum

### PR DESCRIPTION
**Pull request by asubramanya**

### CODE REVIEW

File: libgpos/include/gpos/types.h [hunk 1]
```diff
@@ -96,21 +94,17 @@ namespace gpos
...
-#undef INT_MAX
-#endif // INT_MAX
-#define INT_MAX			((INT) (ULONG_MAX >> 1))
-
-#ifdef INT_MIN
-#undef INT_MIN
-#endif // INT_MIN
-#define INT_MIN			(-INT_MAX - 1)
 
 #ifdef LINT_MAX
```
Comment (hardikar) [inline libgpos/include/gpos/types.h]: Any reason why didn't remove LINT_MAX, LINT_MIN, SINT_MAX & SINT_MIN?

Comment (d) [reply libgpos/include/gpos/types.h]: 1. `INT_MAX`, `INT_MIN`, `ULONG_MAX` and `ULLONG_MAX` are standard macros that we are clobbering, so the minimal fix to enable us to safely `#include` xerces after GPOS headers are to stop clobbering those

1. We could, if we have a consensus, expand the scope to fixing all of those integral limits. I leave it up to the team to decide

**Comment by asubramanya** [comment]
@hardikar Updated the PR to change SINT and LINT MAX and MIN values to enums as well.

File: libgpos/include/gpos/types.h [hunk 2]
```diff
@@ -66,24 +65,23 @@ namespace gpos
 
 	typedef uint32_t ULONG;
```
Comment (melanieplageman) [inline libgpos/include/gpos/types.h]: Could we elaborate on the above comment a little bit? 
- Why is this needed? 
- More importantly, how and when should the API user use it. (i.e. whenever writing code in Orca? only under certain circumstances?)

Also, in the PR or commit message, could you put the reason or motivation for using our own types and for changing them to enums? The namespace prefix is more self-explanatory.

Comment (asubramanya) [reply libgpos/include/gpos/types.h]: I tried looking up the history to find out more about why we created our own types but did not find anything. I will update the commit message with more details from the issue https://github.com/greenplum-db/gporca/issues/358 
before committing it to master.

**Comment by asubramanya** [comment]
@hardikar Corresponding GPDB PR are here: https://github.com/greenplum-db/gpdb/pull/4839

File: libgpos/include/gpos/memory/CCacheAccessor.h [hunk 3]
```diff
@@ -161,9 +161,9 @@ namespace gpos
 				GPOS_ASSERT(NULL == m_pmp);
 
 				// construct a memory pool for cache entry
-				m_pmp = CMemoryPoolManager::Pmpm()->PmpCreate(
-					CMemoryPoolManager::EatTracker, true /*fThreadSafe*/,
-					gpos::ullong_max);
+                m_pmp = CMemoryPoolManager::Pmpm()->PmpCreate(
```
Comment (d) [inline libgpos/include/gpos/memory/CCacheAccessor.h]: use tabs here

File: libgpos/src/memory/CCacheFactory.cpp [hunk 4]
```diff
@@ -70,8 +70,8 @@ CCacheFactory::EresInit()
 	GPOS_RESULT eres = GPOS_OK;
 
 	// create cache factory memory pool
-	IMemoryPool *pmp = CMemoryPoolManager::Pmpm()->PmpCreate(
-		CMemoryPoolManager::EatTracker, true /*fThreadSafe*/, gpos::ullong_max);
+    IMemoryPool *pmp = CMemoryPoolManager::Pmpm()->PmpCreate(
+        CMemoryPoolManager::EatTracker, true /*fThreadSafe*/, gpos::ullong_max);
```
Comment (d) [inline libgpos/src/memory/CCacheFactory.cpp]: why did we change from tab to space here?

File: libnaucrates/src/statistics/CStatsPredUtils.cpp [hunk 5]
```diff
@@ -558,9 +558,9 @@ CStatsPredUtils::AddSupportedStatsFilters
 
 	if (COperator::EopScalarConst == pexprPred->Pop()->Eopid())
 	{
-		pdrgpstatspred->Append(GPOS_NEW(pmp) CStatsPredUnsupported(
-			gpos::ulong_max, CStatsPred::EstatscmptOther,
-			CHistogram::DNeutralScaleFactor));
+        pdrgpstatspred->Append(GPOS_NEW(pmp) CStatsPredUnsupported(
+            gpos::ulong_max, CStatsPred::EstatscmptOther,
```
Comment (d) [inline libnaucrates/src/statistics/CStatsPredUtils.cpp]: I don't think you meant it when you changed the tabs to spaces here.

File: libgpos/include/gpos/types.h [hunk 6]
```diff
@@ -96,41 +94,36 @@ namespace gpos
...
-#ifdef SINT_MAX
-#undef SINT_MAX
-#endif // SINT_MAX
-#define SINT_MAX		((SINT) (USINT_MAX >> 1))
-
-#ifdef SINT_MIN
-#undef SINT_MIN
-#endif // SINT_MIN
-#define SINT_MIN		(-SINT_MAX - 1)
+	enum
```
[the remark this answers is not in the document]
Comment (d) [reply libgpos/include/gpos/types.h]: I doubt it: I vaguely recall the compiler determines the *underlying type* of each enumeration by its enumerators. Which means that you often don't want to mix unsigned enumerators with signed ones -- especially in our case. I need to dig into this a bit more. Sit tight @hardikar

**Comment by d** [comment]
Re-posting just so that it doesn't get buried in the history

@hardikar wrote:
> Just curious - can these be combined into one enum { } declaration?

My take: I doubt it. I vaguely recall the compiler determines the *underlying type* of each enumeration by its enumerators. Which means that you often don't want to mix unsigned enumerators with signed ones -- especially in our case.

The fact that we are using unnamed enumerations doesn't mean that semantically all constants (enumerators) here have the same type. Quite the opposite, you'd generally want `gpos::int_min` to have a different type from `gpos::ulong_max`.

**Comment by d** [comment]
Just talked to @hardikar and @dhanashreek89 . Here's a demonstration of how [mixing enumerators is surprising](https://wandbox.org/permlink/0qB3msgZRqqcgKuH)

```c++
#include <climits>
#include <iostream>

namespace jesse {
enum {
  unnamed_enum_int_min = INT_MIN,
  unnamed_enum_ullong_max = ULLONG_MAX,
};
}

using std::cout;

int main() {
  cout << "unnamed_enum_int_min:\t" << jesse::unnamed_enum_int_min << '\t'
       << "unnamed_enum_ullong_max:\t" << jesse::unnamed_enum_ullong_max
       << '\n';
  return 0;
}

```
